server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -612,4 +634,113 @@ private void failShard(ReplicationFailedException e, IndexShard indexShard) {
+
+        public AsyncFailStaleReplicaTask(IndicesService indicesService, ThreadPool threadPool, PendingCheckpoints pendingCheckpoints) {
+            super(logger, threadPool, INTERVAL, true);
+            this.pendingCheckpoints = pendingCheckpoints;
+            this.indicesService = indicesService;
+            rescheduleIfNecessary();
+        }
+
+        @Override
+        protected boolean mustReschedule() {
+            return true;
+        }
+
+        @Override
+        protected void runInternal() {
+            List<ShardId> shardsToFail = pendingCheckpoints.getStaleShardsToFail();
+            for (ShardId shardId : shardsToFail) {
+                IndexShard replicaShard = this.indicesService.getShardOrNull(shardId);
+                if (replicaShard != null && replicaShard.state() == IndexShardState.STARTED) {
+                    failShard(new ReplicationFailedException("replica too far behind primary, marking as stale"), replicaShard);
+                }
+            }
+        }
+
+        @Override
+        protected String getThreadPool() {
+            return ThreadPool.Names.GENERIC;
+        }
+
+        @Override
+        public String toString() {
+            return "fail_stale_replica";
+        }
+
+    }
+
+    /**
+     * Internal class to track the list of Replication Checkpoints which are still not processed.
+     */
+    public static class PendingCheckpoints {
+        protected final Map<ShardId, List<Tuple<ReplicationCheckpoint, Long>>> checkpointsTracker = ConcurrentCollections
+            .newConcurrentMap();
+        private volatile TimeValue maxAllowedReplicationTime;
+
+        public PendingCheckpoints(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void setMaxAllowedReplicationTime(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void addNewReceivedCheckpoint(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            checkpointsTracker.putIfAbsent(shardId, new ArrayList<>());
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.get(shardId);
+            if (checkpoints.size() > 0) {
+                assert checkpoints.get(checkpoints.size() - 1).v1().isAheadOf(checkpoint) == false;
+            }
+            checkpoints.add(new Tuple<>(checkpoint, System.currentTimeMillis()));
+        }
+
+        public void updateCheckpointProcessed(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            if (checkpointsTracker.containsKey(shardId)) {
+                checkpointsTracker.put(
+                    shardId,
+                    checkpointsTracker.get(shardId)
+                        .stream()
+                        .filter(t -> t.v1().isAheadOf(checkpoint))
+                        .collect(Collectors.toCollection(ArrayList::new))
+                );
```

Comment:
Can this be simplified to `checkpointsTracker.get(shardId).removeIf(t -> checkpoint.isAheadOf(t.v1()));`?

Reply:
Actually, scratch that, I missed that this is a concurrent map. The pattern of check-then-update isn't thread safe. I think it needs to be something like this:

```
checkpointsTracker.computeIfPresent(shardId, (k,v) -> v.stream()
    .filter(t -> t.v1().isAheadOf(checkpoint))
    .collect(Collectors.toCollection(ArrayList::new)));
```